notebook/terminal/__init__.py

```diff
@@ -6,14 +6,20 @@
 if not check_version(terminado.__version__, '0.3.3'):
     raise ImportError("terminado >= 0.3.3 required, found %s" % terminado.__version__)
 
+from ipython_genutils.py3compat import which
 from terminado import NamedTermManager
 from tornado.log import app_log
 from notebook.utils import url_path_join as ujoin
 from .handlers import TerminalHandler, TermSocket
 from . import api_handlers
 
 def initialize(webapp, notebook_dir, connection_url, settings):
-    shell = settings.get('shell_command', [os.environ.get('SHELL') or 'sh'])
+    default_shell = which('sh')
+    if not default_shell and os.name == 'nt':
+        default_shell = 'cmd.exe'
+    shell = settings.get('shell_command',
+        [os.environ.get('SHELL', default_shell)]
```

Comment:
I tend to do defaults for environment variables like `os.environ.get('SHELL') or default_shell`, because setting an environment variable to the empty string is equivalent by convention to unsetting it.